teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,476 @@ While this may be a valuable feature, it should be considered in a separate prop
+It also meets the requirement that users can specify resource requirements for individual `Step`s and `Sidecar`s.
+An alternative option is to use a map of `Step` or `Sidecar` names to container overrides,
+but this violates Kubernetes API convention.
+
+`Step`s and `Sidecar`s are treated separately because they have different fields.
+We may at some point want to override `Step` fields that are not present in `Sidecar`,
+or vice versa. In addition, a `Step` could share a name with a `Sidecar`;
+separating `StepOverrides` and `SidecarOverrides` avoids ambiguity in this case.
+Duplicate names, missing names, or names that don't match `Step` or `Sidecar` names
+will result in the `TaskRun` being rejected.
+
+Some users may want to have resource requirements apply to every `Step` or `Sidecar`,
+or to unnamed `Step`s or `Sidecar`s.
+We could override unnamed `Step`s or `Sidecar`s based on their indices, but we don't currently guarantee
+stable indexing of `Step`s or `Sidecar`s.
+If a `Task` added or removed a `Step` or `Sidecar`, this could break the corresponding `StepOverride`.
+In addition, applying resource requirements to every `Step` or `Sidecar` may not be expected or desirable to users.
+For these reasons, these features will not be supported for the initial version of this proposal.
+We will instead encourage `Task` authors to name their `Step`s and `Sidecar`s.
+This decision may be revisited based on user feedback.
+
+#### Merging resource requirements
+`Step` resource requirements can currently be specified in both
+`Task.Step.Resources` and `Task.StepTemplate.Resources`.
+If resource requirements are specified in both fields, the value present in `Task.Step.Resources`
+is used. However, different resource types (e.g. CPU, memory) are considered independently.
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: image-build-task
+spec:
+  steps:
+    - name: build
+      resources:
+        requests:
+          memory: 500Mi
+        limits:
+          memory: 600Mi
+  stepTemplate:
+    resources:
+      requests:
+        memory: 300Mi
+        cpu: 0.5
+```
+A resulting `TaskRun` will have a memory request of 500Mi, a memory limit of 600Mi,
+and a CPU request of 0.5 CPU units.
+
+This proposal adds a third way to specify `Step` resource requirements: `TaskRun.StepOverrides[].Resources`.
+`TaskRun.StepOverrides[].Resources` will override `Task.Step.Resources` in the same way that that field
+overrides `Task.StepTemplate.Resources`.
+Using the example `Task` defined above, consider the following `TaskRun`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: image-build-taskrun
+spec:
+  taskRef:
+    name: image-build-task
+  stepOverrides:
+    - taskRunStepName: build
+      resources:
+        requests:
+          memory: 1Gi
+```
+The `TaskRun` will have a memory request of 1Gi and no memory limit, as this is the configuration
```

Comment:
> no memory limit

This is kind of surprising, since CPU is being preserved. Why wouldn't it be something like

```
resources:
  requests:
    memory: 1G
    cpu: 0.5
  limits:
    memory: 600Mi
```

Reply:
The override specifies a memory request of 1Gi in this example. it wouldn't make sense for the output container to have a memory request of 1Gi and a memory limit of 600Mi. The stepoverride specifies a memory request which overrides any memory related configuration specified on the step.

Reply:
Good point on the request > limit not making sense.

1. k8s rejects pods at admission that have requests > limits. Instead of bumping the limit if , should we just fail the request?
2. Since there is no mention of the CPU value in the step override, should this value be preserved?

Reply:
1. I updated the TEP to describe the existing behavior of overriding StepTemplates with Steps, since the proposed behavior matches this existing behavior. If the request is > than the limit, tekton will not reject the resulting pod, but k8s will. (maybe we should change this?)
2. CPU value is preserved in this example.